### PR TITLE
Reserves

### LINKED ISSUE #24: Issue with Hydro Pumped Storage values
Hi,

The entsoe pandas client (probably also the Raw client but I didn't tested it), does not take into consideration the "negative values" of the Hydro Pumped Storage. This production is able to pump water and so have a negative production (it's a charge not a generator).

The ENTSOE API website does the difference
image

But the Python client does not :
image

I don't know how to do the difference too in the raw XML
```xml
	<TimeSeries>
		<mRID>5</mRID>
		<businessType>A01</businessType>
		<objectAggregation>A08</objectAggregation>
		<inBiddingZone_Domain.mRID codingScheme="A01">10YFR-RTE------C</inBiddingZone_Domain.mRID>
		<quantity_Measure_Unit.name>MAW</quantity_Measure_Unit.name>
		<curveType>A01</curveType>
		<MktPSRType>
			<psrType>B10</psrType>
		</MktPSRType>
		<Period>
			<timeInterval>
				<start>2019-05-20T05:00Z</start>
				<end>2019-05-20T06:00Z</end>
			</timeInterval>
			<resolution>PT60M</resolution>
				<Point>
					<position>1</position>
                        <quantity>1902</quantity>
				</Point>
		</Period>
	</TimeSeries>
	<TimeSeries>
		<mRID>6</mRID>
		<businessType>A01</businessType>
		<objectAggregation>A08</objectAggregation>
		<outBiddingZone_Domain.mRID codingScheme="A01">10YFR-RTE------C</outBiddingZone_Domain.mRID>
		<quantity_Measure_Unit.name>MAW</quantity_Measure_Unit.name>
		<curveType>A01</curveType>
		<MktPSRType>
			<psrType>B10</psrType>
		</MktPSRType>
		<Period>
			<timeInterval>
				<start>2019-05-20T03:00Z</start>
				<end>2019-05-20T05:00Z</end>
			</timeInterval>
			<resolution>PT60M</resolution>
				<Point>
					<position>1</position>
                        <quantity>1236</quantity>
				</Point>
				<Point>
					<position>2</position>
                        <quantity>308</quantity>
				</Point>
		</Period>
	</TimeSeries>
```

The first TimeSeries should be considerated as positive and the second as negative.

I have send an email to ENTSO-E platform for more information.

### PR DESCRIPTION
**Pull request by jpaduart**
- Implemented queries for reserve (FCR/aFRR/mFRR/RR), both contracted volumes & prices
- A few cosmetic changes
- A fix for the Consumption reporting (fixes #24 ) but also has some side effect on other technologies

### CODE REVIEW

File: entsoe/entsoe.py [hunk 1]
```diff
@@ -428,6 +430,58 @@ def query_imbalance_prices(self, country_code, start, end, psr_type=None):
         response = self.base_request(params=params, start=start, end=end)
         return response.text
 
+    def query_contracted_reserve_prices(self, country_code, start, end,
+                                        type_marketagreement_type, psr_type = None):
```
Comment (JrtPec) [inline entsoe/entsoe.py]: Can you update the docstring for `type_marketagreement_type`?
What values are accepted here? Perhaps a good idea to include them in the docs or `mappings.py`

File: entsoe/entsoe.py [hunk 2]
```diff
@@ -807,6 +861,54 @@ def query_imbalance_prices(self, country_code, start, end, psr_type=None):
...
+        Returns
+        -------
+        pd.DataFrame
+        """
+        text = super(EntsoePandasClient, self).query_contracted_reserve_prices(
+            country_code = country_code, start = start, end = end,
+            type_marketagreement_type = type_marketagreement_type, psr_type = psr_type)
+        df = parse_contracted_reserve(text, TIMEZONE_MAPPINGS[country_code], "procurement_price.amount")
+        df = df.tz_convert(TIMEZONE_MAPPINGS[country_code])
+        return df
```
Comment (JrtPec) [inline entsoe/entsoe.py]: This returns a Time Series, right? Best to add `df = df.truncate(before=start, after=end)`

File: entsoe/parsers.py [hunk 3]
```diff
@@ -73,8 +75,11 @@ def parse_generation(xml_text):
         if series is None:
             all_series[ts.name] = ts
         else:
-            series = series.append(ts)
-            series.sort_index()
+            if (series.name == ts.name) & series.index.equals(ts.index):
+                series = series + ts
```
Comment (JrtPec) [inline entsoe/parsers.py]: Apparently there are cases where, when parsing a generation xml, you encounter a block with the exact same index of the previous block? And then you add them together?
Can you provide some comments/context, perhaps an example so we can verify that this is the correct behaviour?

File: entsoe/parsers.py [hunk 4]
```diff
@@ -346,15 +433,29 @@ def _parse_datetimeindex(soup):
...
     soup : bs4.element.tag
+    tz: str
 
     Returns
     -------
     pd.DatetimeIndex
     """
     start = pd.Timestamp(soup.find('start').text)
     end = pd.Timestamp(soup.find('end').text)
+    if not tz is None:
```
Comment (JrtPec) [inline entsoe/parsers.py]: better form: `if tz is not None`

File: entsoe/parsers.py [hunk 5]
```diff
@@ -346,15 +433,29 @@ def _parse_datetimeindex(soup):
...
+
     delta = _resolution_to_timedelta(res_text=soup.find('resolution').text)
     index = pd.date_range(start=start, end=end, freq=delta, closed='left')
+    if not tz is None:
+        dst_jump = len(set(index.map(lambda d:d.dst()))) > 1
+        if dst_jump and delta == "7D":
+            # For a weekly granularity, if we jump over the DST date in October,
+            # date_range erronously returns an additional index element
+            # because that week contains 169 hours instead of 168.
+            index = index[:-1]
```
Comment (JrtPec) [inline entsoe/parsers.py]: Can you provide an example of a query where this is applied? Because I currently don't see how omitting the last index yields a correct index...
How does this behave in March, when the DST jump adds an extra hour?

File: entsoe/entsoe.py (hunk 1 above)
Comment (jpaduart) [reply entsoe/entsoe.py]: Added definition in mappings.py

File: entsoe/entsoe.py (hunk 2 above)
Comment (jpaduart) [reply entsoe/entsoe.py]: Done

File: entsoe/parsers.py (hunk 3 above)
Comment (jpaduart) [reply entsoe/parsers.py]: This has to do with incorporating consumption (pumped storage, wind off-shore plants...): by adding timeseries with exactly the same index we guarantee that generation and consumption are assimilated.

File: entsoe/parsers.py (hunk 4 above)
Comment (jpaduart) [reply entsoe/parsers.py]: done

File: entsoe/parsers.py (hunk 5 above)
Comment (jpaduart) [reply entsoe/parsers.py]: This one is tricky: remarkably the weekly responses only create an issue for the October DST transition, never in March.

**Comment by jpaduart** [comment]
Made some changes to accommodate for @JrtPec 's remarks.
